### PR TITLE
Improve data locality for statements checker

### PR DESCRIPTION
**Pull request by zintus**
From benchmark on my typical iOS app incremental workload, this change speeds up compiler by 5% +/- 1% (as measured applying this PR to 5.0.1 release)

Bottleneck in my workload is `TypeChecker`. Speedup is achieved by allocating objects representing AST friendlier wrt `TypeChecker` access patterns. 

Purpose of this PR is to better understand:
1\. Are changes towards micro architectural friendliness in specific scenario and CPU desired and could be accepted in future? 

Scenario i wish to improve is working on iOS app on my laptop. I'm daily iterating on iOS app and most of the time it's not optimized, incremental, single file mode compiler invocation. I have specific intel processor at my hands and very specific workload. Do you think specific optimizations worth it?

2\. What's the easiest way to benchmark compiler performance when working on this PR.

I don't think my measurements are good enough because of specific workload/thermals/inability to capture performance counters on macOS. Is it right, that installing Linux and running CI test suite is the way to go?

### CODE REVIEW

**Comment by slavapestov** [comment]
A 5% speedup sounds significant. Let's try compiler performance tests

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Comment by slavapestov** [comment]
(it's going to take a while...)

**Comment by slavapestov** [comment]
Looks like there is a compiler error:

```
/Users/buildnode/jenkins/workspace-private/swift-PR-compiler-performance-macOS/master/new/swift/lib/AST/ASTScopeCreation.cpp:1748:21: error: out-of-line definition of 'operator new' does not match any declaration in 'swift::ast_scope::ASTScopeImpl'
23:26:26 void *ASTScopeImpl::operator new(size_t bytes, const ASTContext &ctx,
23:26:26                     ^~~~~~~~
23:26:26 /Users/buildnode/jenkins/workspace-private/swift-PR-compiler-performance-macOS/master/new/swift/lib/AST/ASTScopeCreation.cpp:1757:17: error: out-of-line definition of 'operator new' does not match any declaration in 'swift::ASTScope'
23:26:26 void *ASTScope::operator new(size_t bytes, const ASTContext &ctx,
23:26:26                 ^~~~~~~~
23:26:26 2 errors generated.
```

Are you able to reproduce it locally?

**Comment by CodaFi** [comment]
I wonder if we can get an even larger win by allocating members into arenas associated with their DeclContexts.  Just sticking all the decls together is going to improve locality by happenstance.

**Comment by zintus** [comment]
Build error should be fixed now. I have also moved all allocations to same arena.

I have tried running benchmark locally, but it fails for me with
```
Benchmark_Driver run: error: argument -i/--independent-samples: expected one argument
```
Are there known workarounds for it?

**Comment by CodaFi** [comment]
The benchmark suite is for the standard library. If you want to run the compiler perf benchmarks, [see Graydon’s instructions](https://github.com/apple/swift/blob/master/docs/CompilerPerformance.md) - but don’t bother with an instruments trace.  You just need the performance counters.

**Comment by CodaFi** [comment]
@swift-ci please test compiler performance

**Comment by CodaFi** [comment]
@swift-ci please test compiler performance

**Comment by CodaFi** [comment]
@swift-ci please test compiler performance

File: lib/AST/Stmt.cpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ using namespace swift;
 // Only allow allocation of Stmts using the allocator in ASTContext.
 void *Stmt::operator new(size_t Bytes, ASTContext &C,
                          unsigned Alignment) {
-  return C.Allocate(Bytes, Alignment);
+  return C.Allocate(Bytes, Alignment, AllocationArena::Declarations);
```
Comment (CodaFi) [inline lib/AST/Stmt.cpp]: Are you sure we want to allocate statements in the arena too?  Seems like it’s going to fragment the arena.

**Comment by CodaFi** [comment]
Huh, these numbers are a little weird, but the important thing seems to be that wall time has gone up quite a bit.  Can you try reverting the last part of the patch and just allocate decls in that arena?

**Comment by zintus** [comment]
Ok, looks like my way to benchmark is inaccurate, I’ll work on reproducing results locally.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by CodaFi** [comment]
@zintus We've made some pretty great strides in the evaluator's memory usage that might necessitate taking a closer look at this. Do you want us to keep this open, or do you want us to close it out so you can open another?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**